### PR TITLE
New resource: azurerm_dedicated_hardware_security_module

### LINKED ISSUE #7602: Support for azurerm_dedicated_hardware_security_module
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

I am working on the implementation of this resource.

### New or Affected Resource(s)

* azurerm_dedicated_hardware_security_module

### Potential Terraform Configuration

```hcl
resource "azurerm_dedicated_hardware_security_module" "test" {
  name                = "acctest-hsm-%s"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name
  sku_name            = "SafeNet Luna Network HSM A790"
  
  network_profile {
    network_interface_private_ip_addresses = ["10.2.1.8"]
	subnet_id                              = azurerm_subnet.test2.id
  }
  
  stamp_id = "stamp2"

  depends_on = [azurerm_virtual_network_gateway.test]
}
```

### References

rest api spec: https://docs.microsoft.com/en-us/cli/azure/ext/hardware-security-modules/?view=azure-cli-latest
Azure doc: https://docs.microsoft.com/en-us/azure/dedicated-hsm/overview
Azure cli: https://docs.microsoft.com/en-us/cli/azure/ext/hardware-security-modules/dedicated-hsm?view=azure-cli-latest#ext-hardware-security-modules-az-dedicated-hsm-create

### PR DESCRIPTION
**Pull request by neil-yechenwei**
Service owner suggests to not use West/North Europe or US East/US East2 for testing because these locations have high customer usage. So I am testing on "canadacentral" location.

(fixes #7602 )

### CODE REVIEW

File: .teamcity/components/generated/services.kt [hunk 1]
```diff
@@ -30,6 +30,7 @@ var services = mapOf(
         "eventgrid" to "EventGrid",
         "eventhub" to "EventHub",
         "frontdoor" to "FrontDoor",
+        "hardwaresecuritymodule" to "Hardware Security Module",
```
Comment (katbyte) [inline .teamcity/components/generated/services.kt:33]: this is a bit long, could we maybe abbreviate it to 
```suggestion
        "hsm" to "Hardware Security Module",
```

File: azurerm/internal/clients/client.go [hunk 2]
```diff
@@ -35,6 +35,7 @@ import (
 	eventgrid "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/eventgrid/client"
 	eventhub "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/eventhub/client"
 	frontdoor "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/frontdoor/client"
+	hardwaresecuritymodules "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/hardwaresecuritymodule/client"
```
Comment (katbyte) [inline azurerm/internal/clients/client.go:38]:
```suggestion
	hsm "github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/hardwaresecuritymodule/client"
```

File: azurerm/internal/clients/client.go [hunk 3]
```diff
@@ -89,80 +90,81 @@ type Client struct {
...
+	DataFactory             *datafactory.Client
+	Datalake                *datalake.Client
+	DataShare               *datashare.Client
+	DevSpace                *devspace.Client
+	DevTestLabs             *devtestlabs.Client
+	Dns                     *dns.Client
+	EventGrid               *eventgrid.Client
+	Eventhub                *eventhub.Client
+	Frontdoor               *frontdoor.Client
+	HardwareSecurityModules *hardwaresecuritymodules.Client
```
Comment (katbyte) [inline azurerm/internal/clients/client.go:122]:
```suggestion
	HSM *hardwaresecuritymodules.Client
```

File: azurerm/internal/clients/client.go [hunk 4]
```diff
@@ -202,6 +204,7 @@ func (client *Client) Build(ctx context.Context, o *common.ClientOptions) error
 	client.EventGrid = eventgrid.NewClient(o)
 	client.Eventhub = eventhub.NewClient(o)
 	client.Frontdoor = frontdoor.NewClient(o)
+	client.HardwareSecurityModules = hardwaresecuritymodules.NewClient(o)
```
Comment (katbyte) [inline azurerm/internal/clients/client.go:207]:
```suggestion
	client.HSM = hsm.NewClient(o)
```

File: azurerm/internal/provider/services.go [hunk 5]
```diff
@@ -31,6 +31,7 @@ import (
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/eventgrid"
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/eventhub"
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/frontdoor"
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/hardwaresecuritymodule"
```
Comment (katbyte) [inline azurerm/internal/provider/services.go:34]:
```suggestion
	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/hsm"
```
and so on

File: azurerm/internal/services/hardwaresecuritymodule/tests/resource_arm_dedicated_hardware_security_module_test.go [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+}
+
+func testAccAzureRMDedicatedHardwareSecurityModule_template(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+provider "azurerm" {
+  features {}
+}
+
+resource "azurerm_resource_group" "test" {
+  name     = "acctest-hsmrg-%d"
```
Comment (katbyte) [inline azurerm/internal/services/hardwaresecuritymodule/tests/resource_arm_dedicated_hardware_security_module_test.go:169]: all rg should start with acctestRG
```suggestion
  name     = "acctestRG-hsm-%d"
```

File: website/docs/r/dedicated_hardware_security_module.html.markdown [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+* `resource_group_name` - (Required) The name of the Resource Group where the Dedicated Hardware Security Module should exist. Changing this forces a new Dedicated Hardware Security Module to be created.
+
+* `location` - (Required) The Azure Region where the Dedicated Hardware Security Module should exist. Changing this forces a new Dedicated Hardware Security Module to be created.
+
+* `sku_name` - (Required) The sku name of the dedicated hardware security module. Changing this forces a new Dedicated Hardware Security Module to be created.
+
+* `network_profile` - (Required)  A `network_profile` block as defined below.
+
+* `stamp_id` - (Required) The ID of the stamp. Changing this forces a new Dedicated Hardware Security Module to be created.
```
Comment (katbyte) [inline website/docs/r/dedicated_hardware_security_module.html.markdown:123]: Could we order these alphabetically o asin schema? 

as well as list the possible values for stand id given there are only two allowed?

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks @neil-yechenwei - overall this looks good, but hardwaresecuritymodule is a bit long, and hsm is a well recognized term. So I think we should use it where possible. But the full term is good for the resource name.

File: .teamcity/components/generated/services.kt (hunk 1 above)
Comment (neil-yechenwei) [reply .teamcity/components/generated/services.kt:33]: updated

File: azurerm/internal/clients/client.go (hunk 2 above)
Comment (neil-yechenwei) [reply azurerm/internal/clients/client.go:38]: updated

File: azurerm/internal/clients/client.go (hunk 3 above)
Comment (neil-yechenwei) [reply azurerm/internal/clients/client.go:122]: updated

File: azurerm/internal/clients/client.go (hunk 4 above)
Comment (neil-yechenwei) [reply azurerm/internal/clients/client.go:207]: updated

File: azurerm/internal/provider/services.go (hunk 5 above)
Comment (neil-yechenwei) [reply azurerm/internal/provider/services.go:34]: updated

File: azurerm/internal/services/hardwaresecuritymodule/tests/resource_arm_dedicated_hardware_security_module_test.go (hunk 6 above)
Comment (neil-yechenwei) [reply azurerm/internal/services/hardwaresecuritymodule/tests/resource_arm_dedicated_hardware_security_module_test.go:169]: updated

File: website/docs/r/dedicated_hardware_security_module.html.markdown (hunk 7 above)
Comment (neil-yechenwei) [reply website/docs/r/dedicated_hardware_security_module.html.markdown:123]: updated

**Comment by neil-yechenwei** [comment]
@katbyte , thanks for your comments. I've updated code.

**Review by katbyte — APPROVED** [review]
Thanks @neil-yechenwei! LGTM

**Pull request merged**